src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,691 @@
+                } else {
+                    resolutionError = AddressResolutionError::TransparentRecipientNotAllowed;
+                }
+            },
+            [&](const SproutPaymentAddress&) {
+                resolutionError = AddressResolutionError::SproutRecipientsNotSupported;
+            },
+            [&](const SaplingPaymentAddress& addr) {
+                if (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxSaplingAvailable) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, addr, payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxSaplingAvailable -= payment.GetAmount();
+                    }
+                } else {
+                    resolutionError = AddressResolutionError::RevealingSaplingAmountNotAllowed;
+                }
+            },
+            [&](const UnifiedAddress& ua) {
+                if (canResolveOrchard
+                    && ua.GetOrchardReceiver().has_value()
+                    && (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxOrchardAvailable)
+                    ) {
+                    resolvedPayments.emplace_back(
+                        ua, ua.GetOrchardReceiver().value(), payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxOrchardAvailable -= payment.GetAmount();
+                    }
+                    orchardOutputs += 1;
+                } else if (ua.GetSaplingReceiver().has_value()
+                    && (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxSaplingAvailable)
+                    ) {
+                    resolvedPayments.emplace_back(
+                        ua, ua.GetSaplingReceiver().value(), payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxSaplingAvailable -= payment.GetAmount();
+                    }
+                } else {
+                    if (strategy.AllowRevealedRecipients()) {
+                        if (ua.GetP2SHReceiver().has_value()) {
+                            resolvedPayments.emplace_back(
+                                ua, ua.GetP2SHReceiver().value(), payment.GetAmount(), std::nullopt, false);
+                        } else if (ua.GetP2PKHReceiver().has_value()) {
+                            resolvedPayments.emplace_back(
+                                ua, ua.GetP2PKHReceiver().value(), payment.GetAmount(), std::nullopt, false);
+                        } else {
+                            // There are no receivers in this UA, which should be impossible.
+                            assert(false);
+                        }
+                    } else if (strategy.AllowRevealedAmounts()) {
+                        resolutionError = AddressResolutionError::TransparentReceiverNotAllowed;
```

Comment:
In the case where we have an Orchard receiver but `canResolveOrchard = false`, and don't have a Sapling receiver, we'll get this error instead of either "insufficient funds" or "can't send to Orchard before NU5". The error message for this is fine, but the documentation of `TransparentReceiverNotAllowed` suggests it is only returned if the UA has no shielded components, which is not necessarily true.